[DataPipe] Fix mishandling of exception message when error is not iterable

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84676

We sometimes get an exception message like this:
```
This exception is thrown by __iter__ of TarArchiveLoaderIterDataPipe(datapipe=FileOpenerIterDataPipe, length=-1, mode='r:')    elif msg not in e.args[0] and single_iterator_msg not in e.args[0]:

TypeError: argument of type 'int' is not iterable
```

The `TypeError` raised by the mishandling of the error message obfuscates the true exception, which now will be show as:
```
FileNotFoundError: [Errno 2] No such file or directory:
```

NivekT committed:
[DataPipe] Fix mishandling of exception message when error is not iterable

[ghstack-poisoned]

NivekT commented:
@pytorchbot rebase -s

NivekT committed:
Update on "[DataPipe] Fix mishandling of exception message when error is not iterable"


We sometimes get an exception message like this:
```
This exception is thrown by __iter__ of TarArchiveLoaderIterDataPipe(datapipe=FileOpenerIterDataPipe, length=-1, mode='r:')    elif msg not in e.args[0] and single_iterator_msg not in e.args[0]:

TypeError: argument of type 'int' is not iterable
```

The `TypeError` raised by the mishandling of the error message obfuscates the true exception, which now will be show as:
```
FileNotFoundError: [Errno 2] No such file or directory:
```

[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/_hook_iterator.py:
```diff
@@ -195,8 +195,13 @@ def wrap_generator(*args, **kwargs):
                     full_msg = f"{msg} {datapipe.__class__.__name__}({_generate_input_args_string(datapipe)})"
                     if len(e.args) == 0:  # If an exception message doesn't exist
                         e.args = (f'\nThis exception is {full_msg}',)
-                    elif msg not in e.args[0] and single_iterator_msg not in e.args[0]:
```
IIUC the Error is raised due to `msg not in e.args[0]` where `e.args[0]` is an integer. Then why not we just do it:
```py
elif not isinstance(e.args[0], str):
    ...
elif msg not in e.args[0] and single_iterator_msg not in e.args[0]:
    ...
```

NivekT replied:
Yea that's definitely better. Thaks!

NivekT committed:
Update on "[DataPipe] Fix mishandling of exception message when error is not iterable"


We sometimes get an exception message like this:
```
This exception is thrown by __iter__ of TarArchiveLoaderIterDataPipe(datapipe=FileOpenerIterDataPipe, length=-1, mode='r:')    elif msg not in e.args[0] and single_iterator_msg not in e.args[0]:

TypeError: argument of type 'int' is not iterable
```

The `TypeError` raised by the mishandling of the error message obfuscates the true exception, which now will be show as:
```
FileNotFoundError: [Errno 2] No such file or directory:
```

[ghstack-poisoned]

NivekT commented:
@pytorchbot rebase -s

NivekT committed:
Update on "[DataPipe] Fix mishandling of exception message when error is not iterable"


We sometimes get an exception message like this:
```
This exception is thrown by __iter__ of TarArchiveLoaderIterDataPipe(datapipe=FileOpenerIterDataPipe, length=-1, mode='r:')    elif msg not in e.args[0] and single_iterator_msg not in e.args[0]:

TypeError: argument of type 'int' is not iterable
```

The `TypeError` raised by the mishandling of the error message obfuscates the true exception, which now will be show as:
```
FileNotFoundError: [Errno 2] No such file or directory:
```

[ghstack-poisoned]

ejguan reviewed (approved):
LGTM
